### ISSUE TITLE
Several issues with saving model by averaging multiple models

### ISSUE THREAD

**Issue by sun-peach** [tensorflow/tensorflow#33792]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 1.13
- Python version: 2.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10
- GPU model and memory: 2080 ti

I try to average two models weights to form one final model. The two original models are initialized following the same setting up. The only difference is they are trained with different data. After train them, I load them and average them using the code below:

```
import tensorflow as tf
import os
sess1 = tf.Session()
sess2 = tf.Session()
# toy example, so load the same model twice and do averaging
saver1 = tf.train.import_meta_graph('model.meta')
saver2 = tf.train.import_meta_graph('model.meta')  
saver1.restore(sess1, 'model')
saver2.restore(sess2, 'model')
with tf.Session() as sess3:
    sess3.run(init_op)
    all_vars = tf.trainable_variables()
    values1 = sess1.run(all_vars)
    values2 = sess2.run(all_vars)
    all_assign = []
    for var, val1, val2 in zip(all_vars, values1, values2):
        all_assign.append(tf.assign(var, (val1 + val2)/ 2))
    sess3.run(all_assign)
    saver3 = tf.train.Saver()
    save_path = saver3.save(sess3, os.path.join('./debug/', 'model'))
```
Then it shows the error:
At least two variables have the same name: beta1_power

I have two questions:
1). is this way I do averaging right?
2). why I have the error?

Thank you very much!

**Comment by ravikyram** [comment]
@sun-peach 
Please, let us know what error you are getting.Can you share the simple standalone code to reproduce the issue in our environment. It helps in localizing the issue faster. Thanks!

**Comment by sun-peach** [comment]
@ravikyram hi, I have solved the second question. But for the first question, I still don't know how to average two models. I tried a little bit, but result is not right.

I split the data to 2 partition and train as 2 jobs, resulting in 2 models. Eventually, I average the 2 models. But, the resulting model has validation accuracy 0.0. Previously, when I train the model with all data and without partitioning and averaging, the accuracy can approach 98%.

**Comment by sun-peach** [comment]
The current code is below:
```
with tf.Graph().as_default():
        # define the main net
        nnets_to_average = nnets_list.split(" ")  # list of nnets to be averaged in a string delimited by space 
        main_net_dir = nnets_to_average[0]
        # meta file from main net (assuming all nets have exactly same architecture)
        saver = tf.train.import_meta_graph(os.path.join(main_net_dir,'model.meta'))
        all_vars = tf.trainable_variables()
        # read all model params
        model_params = []
        for net_dir in nnets_to_average:
            temp_sess = tf.Session()
            saver.restore(temp_sess, os.path.join(net_dir,'model'))
            temp_param = temp_sess.run(all_vars)
            model_params.append(temp_param)
            temp_sess.close()
        # do averaging
        #init_op = tf.global_variables_initializer()
        sess_final = tf.Session()
        saver.restore(sess_final, os.path.join(main_net_dir,'model'))       # main net as a template
        all_assign = []
        sum_count = 1
        for var_list in zip(all_vars, *model_params):
            var_list = list(var_list)
            var = var_list[0]
            temp_sum = var_list[1]
            if len(var_list)>=3:
                for sub_var in var_list[2:]:
                    temp_sum += sub_var
                    sum_count += 1
            all_assign.append(tf.assign(var, temp_sum/sum_count))
        sess_final.run(all_assign)
        # save averaged model and copy remaining file
        out_dir = '{0}/model_{1}'.format(dir, iter+1)
        if not os.path.exists(out_dir):
            os.makedirs(out_dir)
        saver.save(sess_final, os.path.join(out_dir,'model'))
        sess_final.close()
```

**Comment by jvishnuvardhan** [comment]
This is not Build/Installation or Bug/Performance issue. Please post this kind of support questions at [Stackoverflow](https://stackoverflow.com/questions/tagged/tensorflow). There is a big community to support and learn from your questions. GitHub is mainly for addressing bugs in installation and performance. Thanks
